### ISSUE TITLE
The loss of D in the code is different from that in the paper

### ISSUE THREAD

**Issue by EvaFlower** [lfovia/QAGANS#3]
Hi there,

I found the d_loss defined in the code is different: 	https://github.com/lfovia/QAGANS/blob/184b57cb79d9331387b5e8c3b8c2268ca72d9cf9/QAGAN-SSIM/celebA_SSIM_GP.py#L248
which is instead d_loss = wasserstein + regularizer in the paper Eq. (13).

**Comment by lfovia** [comment]
Hi ,
The main regulariser  for our  QAGAN-SSIM is WGAN-GP+SSIM penalty.
The extra stabiliy regulariser was added  in the code to stop it from drifting from during training. It was motivated by previous methods [1],[2]. The stability factor was in very less proportion compared to the main regularisers.
1.Banach WGAN  https://arxiv.org/pdf/1806.06621.pdf  
2. PGGAN  https://arxiv.org/pdf/1710.10196.pdf
Hope this clarifies your question.  Thanks.

**Comment by lfovia** [comment]
Hi, 

If i understand your question correctly. You are asking about the negative sign in the loss term. 
We can define the Wasserstein GAN loss in both ways.

in our case:
dloss = -wasserstein + regulariser.
gloss = tf.reduce_mean(d_generated_train) .

Other case :
dloss = wasserstein + regulariser
gloss = -tf.reduce_mean(d_generated_train) 

Losses are going to be minimised with respect to variables in both the cases.  These two methods are equivalent. We followed the first one. Thanks.

**Comment by EvaFlower** [comment]
Hi, 
Thanks for your clarification of the added regulariser. The second difference is that you use minus wassersetein (-wasserstein) in the code, instead wasserstin. Could you help clarify this also?
Thanks.

**Comment by lfovia** [comment]
Hi,

If i understand your question correctly. You are asking about the negative sign in the loss term.
We can define the Wasserstein GAN loss in both ways.

in our case:
dloss = -wasserstein + regulariser.
gloss = tf.reduce_mean(d_generated_train) .

Other case :
dloss = wasserstein + regulariser
gloss = -tf.reduce_mean(d_generated_train)

Losses are going to be minimised with respect to variables in both the cases. These two methods are equivalent. We followed the first one. Thanks.

**Comment by EvaFlower** [comment]
Yes, your understanding is right. Now it is clear to me. Thank you very much.

Other questions about your implementation:
1. Why you use tf.image.image_to_grayscale when calculating the two quality regularizer?
2. Since there is a division when calculating SSIM_penalty, if I understand correctly, 0.01 in the following code is to avoid the numerical instability, is it right?
https://github.com/lfovia/QAGANS/blob/184b57cb79d9331387b5e8c3b8c2268ca72d9cf9/QAGAN-SSIM/SSIM_penalty.py#L50
May we also need to add it for
https://github.com/lfovia/QAGANS/blob/184b57cb79d9331387b5e8c3b8c2268ca72d9cf9/QAGAN-SSIM/SSIM_penalty.py#L49
Right? Thanks.

**Comment by lfovia** [comment]
Hi,

1. SSIM in it's standard form is defined on the grayscale images. So we convert the images from 
rgb2gray. We used grayscale SSIM.
2. Yes, You are correct.  The epsilon is added to denominator to avoid numerical stability issues. You can add to S1 also. In our experiments, we found that S2 is more prone to stability issues.
Thanks

**Comment by EvaFlower** [comment]
Thanks a lot for all your detailed explanation.
Regards,
